x/merkledb/db.go

```diff
@@ -460,8 +461,26 @@ func (db *merkleDB) Close() error {
 		return err
 	}
 
-	// Successfully wrote intermediate nodes.
-	return db.baseDB.Put(cleanShutdownKey, hadCleanShutdown)
+	var (
+		batch = db.baseDB.NewBatch()
+		err   error
+	)
+	// Write the root key
+	if db.root.IsNothing() {
+		err = batch.Delete(rootDBKey)
+	} else {
+		rootKey := encodeKey(db.root.Value().key)
+		err = batch.Put(rootDBKey, rootKey)
+	}
+	if err != nil {
+		return err
+	}
```

Comment:
We must write the root key here to avoid the case where:
1. An empty database is corrupted.
2. Repairing an empty database doesn't update the root key
3. Closing the now (partially) repaired database updates the clean shutdown flag (but leaves the root key as invalid).